core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -277,11 +287,21 @@ public static TableMetadata read(FileIO io, String path) {
 
   public static TableMetadata read(FileIO io, InputFile file) {
     Codec codec = Codec.fromFileName(file.location());
-    try (InputStream is = file.newStream();
-        InputStream gis = codec == Codec.GZIP ? new GZIPInputStream(is) : is) {
+    InputStream is = file.newStream();
+    // if codec is GZIP, is will be closed by GZIPInputStream
+    // otherwise, os will be closed by try-with-resources
+    try (InputStream gis = codec == Codec.GZIP ? new GZIPInputStream(is) : is) {
```

Comment:
Why not gathering both statements in one try close ? It would be easier.

Something like:

```
try (InputStream is = (codec == Codec.GZIP) ? new GzipInputStream(file.newStream()) : file.newStream()) {
```

Reply:
This was the original code, but there is an edge case where the `GzipInputStream` throws an exception in its constructor, and it does not close the InputStream passed as an argument in this case, which according to the PR that introduced the bug could increase memory usage and lead to unexpected results.